changefeed: fix race in cloudstorage sink

asg0451 opened the pull request:
When using zstd or "fast" gzip compression, the
cloudstorage sink was susceptible to a data race
on its buffered data. This was due to unexpected
asynchronousness in the compression writers.


Fixes: #130656
Fixes: #130651


Release note (bug fix): Fix a data race in the cloudstorage sink.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/146297)
<!-- Reviewable:end -->

asg0451 committed:
changefeed: fix race in cloudstorage sink

When using zstd or "fast" gzip compression, the
cloudstorage sink was susceptible to a data race
on its buffered data. This was due to unexpected
asynchronousness in the compression writers.

Fixes: #130656
Fixes: #130651

Release note (bug fix): Fix a data race in the cloudstorage sink.

rharding6373 commented on pkg/ccl/changefeedccl/sink_cloudstorage.go:
```diff
@@ -102,9 +107,19 @@ var _ io.Writer = &cloudStorageSinkFile{}
 	f.rawSize += len(p)
 	if f.codec != nil {
-		return f.codec.Write(p)
+		n, err := f.codec.Write(p) // WRITE
```
nit: left a debugging comment

rharding6373 reviewed (approved):
LGTM, nice fix! Only a couple nits

andyyang890 commented on pkg/ccl/changefeedccl/sink_cloudstorage.go:
```diff
@@ -68,11 +69,15 @@ func cloudStorageFormatTime(ts hlc.Timestamp) string {
+	// bufLen tracks the length of buf. Because some of the codecs write to the
+	// buffer asynchronously, we can't safely call `buf.Len()` while the codec
+	// is open.
+	bufLen        int64
```
How awful do you think the performance would be if we wrapped the buffer with a mutex for just the pgzip writer case? I'm a little worried we might forget to update `bufLen` somewhere in the future.

andyyang890 commented on pkg/ccl/changefeedccl/changefeed_test.go:
```diff
@@ -11573,3 +11573,46 @@ func TestChangefeedMVCCTimestampWithQueries(t *testing.T) {
+func TestCloudstorageParallelCompression(t *testing.T) {
+	defer leaktest.AfterTest(t)()
+	defer log.Scope(t).Close(t)
+	skip.UnlessUnderRace(t)
```
optional super nit: add a newline before this

asg0451 commented:
bors r=andyyang890
